src/coprocessor/dag/rpn_expr/types.rs

```diff
@@ -0,0 +1,695 @@
+        use std::convert::{TryFrom, TryInto};
+
+        let field_type = def.take_field_type();
+        let eval_type =
+            EvalType::try_from(field_type.tp()).map_err(|e| Error::Other(box_err!(e)))?;
+
+        match def.get_tp() {
+            ExprType::Null => {
+                let scalar_value = match eval_type {
+                    EvalType::Int => ScalarValue::Int(None),
+                    EvalType::Real => ScalarValue::Real(None),
+                    EvalType::Decimal => ScalarValue::Decimal(None),
+                    EvalType::Bytes => ScalarValue::Bytes(None),
+                    EvalType::DateTime => ScalarValue::DateTime(None),
+                    EvalType::Duration => ScalarValue::Duration(None),
+                    EvalType::Json => ScalarValue::Json(None),
+                };
+                rpn_nodes.push(RpnExpressionNode::Constant {
+                    value: scalar_value,
+                    field_type,
+                });
+            }
+            ExprType::Int64 => {
+                let scalar_value = match eval_type {
+                    EvalType::Int => {
+                        let value = number::decode_i64(&mut def.get_val()).map_err(|_| {
+                            Error::Other(box_err!(
+                                "Unable to decode {:?} from the request",
+                                eval_type
+                            ))
+                        })?;
+                        ScalarValue::Int(Some(value))
+                    }
+                    t => {
+                        return Err(box_err!(
+                            "Unexpected eval type {:?} for ExprType {:?}",
+                            t,
+                            def.get_tp()
+                        ));
+                    }
+                };
+                rpn_nodes.push(RpnExpressionNode::Constant {
+                    value: scalar_value,
+                    field_type,
+                });
+            }
+            ExprType::Uint64 => {
+                let scalar_value = match eval_type {
+                    EvalType::Int => {
+                        let value = number::decode_u64(&mut def.get_val()).map_err(|_| {
+                            Error::Other(box_err!(
+                                "Unable to decode {:?} from the request",
+                                eval_type
+                            ))
+                        })?;
+                        ScalarValue::Int(Some(value as i64))
+                    }
+                    t => {
+                        return Err(box_err!(
+                            "Unexpected eval type {:?} for ExprType {:?}",
+                            t,
+                            def.get_tp()
+                        ));
+                    }
+                };
+                rpn_nodes.push(RpnExpressionNode::Constant {
+                    value: scalar_value,
+                    field_type,
```

Comment:
It's repeated in each match case, could we do it after the `match`

Reply:
Of course not XD Because we have 3 types to push, Constant, Fn or ColumnRef.